ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java

```diff
@@ -506,10 +478,23 @@ protected Boolean findNextCompactionAndExecute(boolean computeStats) {
         task (currently we're using Tez split grouping).
         */
         QueryCompactor queryCompactor = QueryCompactorFactory.getQueryCompactor(t, conf, ci);
+        boolean generic_stats = conf.getBoolVar(HiveConf.ConfVars.HIVE_COMPACTOR_GATHER_STATS);
+        boolean mr_stats = conf.getBoolVar(HiveConf.ConfVars.HIVE_MR_COMPACTOR_GATHER_STATS);
+        boolean computeStats = (queryCompactor == null && (mr_stats || generic_stats)) || generic_stats;
+
+        LOG.info("Starting " + ci.type.toString() + " compaction for " + ci.getFullPartitionName() + " in " +
+                compactionTxn + " with compute stats set to " + computeStats);
+
+        StatsUpdater su = null;
+        if (computeStats) {
+          su = StatsUpdater.init(ci, conf, runJobAsSelf(ci.runAs) ? ci.runAs : t.getOwner(),
```

Comment:
Q. can we refactor StatsUpdater to a service utility that just submits the compute stats query and not generate a separate instance per each compaction candidate?